File: src/main/java/net/katsstuff/danmakucore/entity/spellcard/Spellcard.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity.spellcard;

import javax.annotation.Nullable;

import net.katsstuff.danmakucore.DanmakuCore;
import net.katsstuff.danmakucore.entity.living.TouhouCharacter;
import net.katsstuff.danmakucore.registry.RegistryValueItemStack;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;
import net.minecraft.util.ResourceLocation;
import net.minecraft.world.World;

@SuppressWarnings("UnusedParameters")
public abstract class Spellcard extends RegistryValueItemStack<Spellcard> {

	public Spellcard(String name) {
		setRegistryName(name);
	}

	public Spellcard() {}

	public final void bakeModel() {
		DanmakuCore.proxy.bakeSpellcard(this);
	}

	public abstract SpellcardEntity instantiate(EntitySpellcard card, @Nullable EntityLivingBase target);

	/**
	 * @return The needed bombs to execute the spellcard.
	 */
	public abstract int getLevel();

	/**
	 * @return The remove time. How long the spellcard will remove danmaku for.
	 */
	public abstract int getRemoveTime();

	/**
	 * @return The end time. How long the spellcard will physically exist and do stuff. Can also be
	 * though of as cooldown.
	 */
	public abstract int getEndTime();

	/**
	 * @return The Touhou character that uses this spellcard in the games.
	 */
	public abstract TouhouCharacter getOriginalUser();

	/**
	 * Called when a spellcard itemStack is rightclicked.
	 *
	 * @return If the spellcard should be declared.
	 */
	@Override
	public boolean onRightClick(ItemStack stack, World world, EntityPlayer player, EnumHand hand) {
		return true;
	}

	/**
	 * Called before declaring a spellcard of this type.
	 *
	 * @return If the spellcard is allowed to be declared.
	 */
	public boolean onDeclare(EntityLivingBase user, @Nullable EntityLivingBase target, boolean firstAttack) {
		return true;
	}

	@Override
	public int compareTo(Spellcard other) {
		return getOriginalUser().getFullName().compareToIgnoreCase(other.getOriginalUser().getFullName());
	}

	@Override
	public String getUnlocalizedName() {
		return "spellcard." + getModId() + "." + getName();
	}

	@Override
	public ModelResourceLocation getItemModel() {
		return new ModelResourceLocation(new ResourceLocation(getModId(), "danmaku/spellcard/" + getName()), "inventory");
	}
}

File: src/main/scala/net/katsstuff/danmakucore/entity/spellcard/Spellcard.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.entity.spellcard

import javax.annotation.Nullable

import net.katsstuff.danmakucore.DanmakuCore
import net.katsstuff.danmakucore.entity.living.TouhouCharacter
import net.katsstuff.danmakucore.registry.RegistryValueItemCreatable
import net.katsstuff.danmakucore.scalastuff.TouhouHelper
import net.katsstuff.teamnightclipse.mirror.data.Vector3
import net.minecraft.client.renderer.block.model.{ModelResourceLocation => MRL}
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.{EntityLiving, EntityLivingBase}
import net.minecraft.util.{EnumHand, ResourceLocation}
import net.minecraft.world.World

abstract class Spellcard extends RegistryValueItemCreatable[Spellcard, EntitySpellcard] {

  def this(name: String) {
    this()
    setRegistryName(name)
    DanmakuCore.proxy.bakeSpellcard(this)
  }

  def instantiate(card: EntitySpellcard, @Nullable target: EntityLivingBase): SpellcardEntity =
    instantiate(card, Option(target))

  /**
    * Creates an instance of this spellcard type.
    */
  def instantiate(card: EntitySpellcard, target: Option[EntityLivingBase]): SpellcardEntity

  /**
    * The needed bombs to execute the spellcard.
    */
  def level: Int

  /**
    * The remove time. How long the spellcard will remove danmaku for.
    */
  //noinspection MutatorLikeMethodIsParameterless
  def removeTime: Int

  /**
    * The end time. How long the spellcard will physically exist and do stuff. Can also be
    * though of as cooldown.
    */
  def endTime: Int

  /**
    * The Touhou character that uses this spellcard in the games.
    */
  def touhouUser: TouhouCharacter

  /**
    * Creates and spawns an entity when a player uses this spellcard.
    */
  def create(player: EntityPlayer, firstAttack: Boolean): Option[EntitySpellcard] =
    TouhouHelper.declareSpellcardPlayer(player, this, firstAttack)

  override def create(
      world: World,
      optUser: Option[EntityLivingBase],
      alternateMode: Boolean,
      pos: Vector3,
      direction: Vector3,
      hand: Option[EnumHand]
  ): Option[EntitySpellcard] = {
    optUser.flatMap { user =>
      user match {
        case entityPlayer: EntityPlayer => TouhouHelper.declareSpellcardPlayer(entityPlayer, this, firstAttack = true)
        case _ =>
          val target = user match {
            case living: EntityLiving if living.getAttackTarget != null => living.getAttackTarget
            case _                                                      => user.getRevengeTarget
          }

          TouhouHelper.declareSpellcard(user, Option.apply(target), this, firstAttack = false, addSpellcardName = true)
      }
    }
  }

  /**
    * Called before declaring a spellcard of this type.
    *
    * @return If the spellcard is allowed to be declared.
    */
  def beforeDeclare(user: EntityLivingBase, @Nullable target: EntityLivingBase, firstAttack: Boolean) = true

  override def unlocalizedName: String = s"spellcard.$modId.$name"

  override def itemModel: MRL = new MRL(new ResourceLocation(modId, s"danmaku/spellcard/$name"), "inventory")
}
object Spellcard {
  implicit val ordering: Ordering[Spellcard] = Ordering.by((card: Spellcard) => card.fullNameString)
}
